feat(analytics): Dart implementation

fjnoyp opened the pull request:
*Issue #, if available:*

*Description of changes:*

Merge analytics in next release branch 

All the changes here have already been reviewed and approved.  

This is the main analytics dart first PR + db_common refactor PR. 

Reference: 

Analytics dart
https://github.com/aws-amplify/amplify-flutter/pull/2197

db_common
https://github.com/aws-amplify/amplify-flutter/pull/2316


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ragingsquirrel3 commented:
request: Could CI errors be fixed here where possible? Specifically flutter unit test for amplify_flutter https://circleci.com/gh/aws-amplify/amplify-flutter/71220 and format https://circleci.com/gh/aws-amplify/amplify-flutter/71212?

ragingsquirrel3 commented on packages/analytics/amplify_analytics_pinpoint/example/web/index.html:
```diff
@@ -0,0 +1,16 @@
+
+<body>
+<!-- Added for Drift web setup: https://drift.simonbinder.eu/web/ -->
+<script defer src="sql-wasm.js"></script>
```
Likewise, does it need the wasm file itself sibling here? sql-wasm.wasm

fjnoyp commented:
Thanks @ragingsquirrel3 I ran a format on the analytics dart file.  Unsure on how unit tests are failing as Analytics doesn't actually introduce any!  I can check more into that.

ragingsquirrel3 commented:
Why was a license file moved to /tool dir? at top-level of repo.

ragingsquirrel3 commented on packages/amplify/amplify_flutter/lib/src/method_channel_amplify.dart:
```diff
@@ -126,13 +119,4 @@ class MethodChannelAmplify extends AmplifyClassImpl {
       }
     }
   }
-
```
question: why was this part removed here?

ragingsquirrel3 replied:
talked offline, seems removed bc superclass does same thing

fjnoyp commented:
I think the LICENSE was added by some build step on accident, I removed it.  

For that line I think if you look at the entire file you'll see that I deleted the reset method that space was there to separate the methods but just below is another closing bracket.

ragingsquirrel3 commented on packages/analytics/amplify_analytics_pinpoint/pubspec.yaml:
```diff
@@ -9,26 +9,28 @@ environment:
-  amplify_core: ">=1.0.0-next.0+2 <1.0.0-next.1"
-  aws_common: ">=0.2.4 <0.3.0"
+  amplify_analytics_pinpoint_dart: ^0.1.0
+  amplify_core: ">=1.0.0-next.0+1 <1.0.0-next.1"
```
question: Are we sure these deps are correct here from last rebase? seems like other packages get core from .0+2, also noticing the ways of specifying below deps like amplify_secure_storage different than auth.

dnys1 replied:
They will be updated in version bump - no worries if they're stale here.

ragingsquirrel3 commented on packages/analytics/amplify_analytics_pinpoint_dart/pubspec.yaml:
```diff
@@ -0,0 +1,38 @@
+description: A Dart-only implementation of the Amplify Analytics plugin for Pinpoint.
+version: 0.1.0
+homepage: https://docs.amplify.aws/lib/q/platform/flutter/
+repository:
```
nit: repo field blank

ragingsquirrel3 commented on packages/analytics/amplify_analytics_pinpoint_dart/pubspec.yaml:
```diff
@@ -0,0 +1,38 @@
+  sdk: ">=2.17.0 <3.0.0"
+
+dependencies:
+  amplify_core: ^1.0.0-next.0
```
question: Should these deps be referred in similar manner as other plugins?

fjnoyp replied:
I think we can disregard like Dillon said and commit to changing this in the version bump PR

dnys1 commented on .github/workflows/lint.yml:
```diff
@@ -172,6 +172,58 @@ jobs:
         if: "always() && steps.packages_common_amplify_db_common_example_pub_upgrade.conclusion == 'success'"
         working-directory: packages/common/amplify_db_common/example
         run: flutter analyze --fatal-infos lib
+      - id: packages_storage_amplify_storage_s3_example_linux_flutter_ephemeral_.plugin_symlinks_amplify_db_common_pub_upgrade
```
Whoops

fjnoyp replied:
Thanks reran the mono_repo generate

ragingsquirrel3 commented on packages/amplify/amplify_flutter/test/amplify_test.dart:
```diff
@@ -13,7 +13,7 @@
  */
 
-import 'package:amplify_analytics_pinpoint/method_channel_amplify.dart';
+import 'package:amplify_datastore/method_channel_datastore.dart';
```
This test is still failing, will ping offline w suggestions.

fjnoyp replied:
Thanks for the fixes

ragingsquirrel3 reviewed (approved):
Still some lint errors looks like in CI? Would be good to fix those, seems like in amplify_analytics_pinpoint_dart and just a few things. Otherwise looks good from a sanity check and checking integration of latest fixes as I didn't completely re-review the entire analytics implementation that was already reviewed/approved.
